### PR TITLE
merge master into config flow. config flow : tested

### PR DESCRIPTION
**Pull request by BlueBlueBlob**
Tested :)
And super fancy

### CODE REVIEW

**Comment by BlueBlueBlob** [comment]
I have something big incoming. I will post some screenshots ;)

**Comment by BlueBlueBlob** [comment]
Enjoy :)

File: README.md [hunk 1]
```diff
@@ -6,20 +6,10 @@
 4. Restart Home Assistant
 5. Go to Grocy-Wrench icon-Manage API keys-Add
 6. Copy resulting API key and Grocy URL and input this in configuration.yaml:
+7. Choose:
+   - Add `Grocy:` to your HA configuration.
```
Comment (SebRut) [inline README.md]: I guess this should be `grocy:`

File: custom_components/grocy/.translations/en.json [hunk 2]
```diff
@@ -0,0 +1,23 @@
+    {
+        "config": {
+            "title": "Grocy",
+            "step": {
+                "user": {
+                    "title": "Grocy",
+                    "description": "If you need help with the configuration have a look here: https://github.com/custom-components/grocy",
+                    "data": {
+                        "url": "Grocy API URL (e.g. \"http://yourgrocyurl.com\"",
```
Comment (SebRut) [inline custom_components/grocy/.translations/en.json]: closing bracket is missing

File: custom_components/grocy/__init__.py [hunk 3]
```diff
@@ -31,9 +35,15 @@
...
+    STOCK_NAME,
+    CHORES_NAME,
+    SHOPPING_LIST_NAME,
+    EXPIRING_PRODUCTS_NAME,
+    EXPIRED_PRODUCTS_NAME,
+    MISSING_PRODUCTS_NAME,
 )
 
-MIN_TIME_BETWEEN_UPDATES = timedelta(seconds=60)
+MIN_TIME_BETWEEN_UPDATES = timedelta(seconds=300)
```
Comment (SebRut) [inline custom_components/grocy/__init__.py]: Is this related to adding config flow?

File: custom_components/grocy/__init__.py [hunk 4]
```diff
@@ -216,3 +260,16 @@ def __init__(self, hass, client):
...
 
     return returnvalue
+
+async def async_remove_entry(hass, config_entry):
+    """Handle removal of an entry."""
+    try:
+        await hass.config_entries.async_forward_entry_unload(config_entry, "sensor")
+        _LOGGER.info("Successfully removed sensor from the grocy integration")
+    except ValueError:
+        pass
```
Comment (SebRut) [inline custom_components/grocy/__init__.py]: Could you add error logging please?

File: custom_components/grocy/binary_sensor.py [hunk 5]
```diff
@@ -13,41 +14,60 @@
...
         self.attr["items"] = jsonpickle.encode(
-            expiring_products,
+            data,
             unpicklable=False)
 
+    @property
+    def unique_id(self):
+        """Return a unique ID to use for this binary_sensor."""
+        return (
+            "e308a906-4c3f-4a58-a067-47d04d3e196e" + self.name
```
Comment (SebRut) [inline custom_components/grocy/binary_sensor.py]: is this the right way to specify an unique_id?

File: custom_components/grocy/config_flow.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+        if user_input is not None:
+            valid = await self._test_credentials(
+                user_input["url"], user_input["api_key"],
+                user_input["port"],user_input["verify_ssl"]
+            )
+            if valid:
+                return self.async_create_entry(title="Grocy", data=user_input)
+            else:
+                self._errors["base"] = "auth"
```
Comment (SebRut) [inline custom_components/grocy/config_flow.py]: Could you add error logging here?

File: custom_components/grocy/config_flow.py [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
+        return self.async_create_entry(title="configuration.yaml", data={})
+
+    async def _test_credentials(self, url, api_key, port, verify_ssl):
+        """Return true if credentials is valid."""
+        try:
+            client = Grocy(url, api_key, port, verify_ssl)
+            client.stock()
+            return True
+        except Exception:  # pylint: disable=broad-except
```
Comment (SebRut) [inline custom_components/grocy/config_flow.py]: error logging please

File: custom_components/grocy/__init__.py (hunk 3 above)
Comment (BlueBlueBlob) [reply custom_components/grocy/__init__.py]: Yes and no. Since add-on is not up to data, we can not check if grocy db was updated.
With 30 seconds and all sensors, grocy add-on cpu usage is high asf...

File: README.md (hunk 1 above)
Comment (BlueBlueBlob) [reply README.md]: Yes, sorry this is the component name and not showing name.

File: custom_components/grocy/binary_sensor.py (hunk 5 above)
Comment (SebRut) [reply custom_components/grocy/binary_sensor.py]: I'd be hesistant to use a generated uuid for our sensors, not sure of grocy provides some kind of id. Otherwise we have to use various parameters and maybe hash them or something to generate a uuid. Maybe we could also replace the uuid with an (md5?) hash of the api key. This would ensure stable identifiers as long the api key doesn'T change.

Comment (BlueBlueBlob) [reply custom_components/grocy/binary_sensor.py]: https://developers.home-assistant.io/docs/en/entity_registry_index.html

Comment (BlueBlueBlob) [reply custom_components/grocy/binary_sensor.py]: I will try your method and show result.
If we have duplicate sensors (*_2) : not a good id.
Keep in touch :)

**Comment by BlueBlueBlob** [comment]
I think config_flow is done.
Works nice for me. Just we need an add-on update, because sensors take to much time to update.
Or optimize pygrocy lib. (No idea for now)

**Comment by BlueBlueBlob** [comment]
My core.entity_registry from HA : 
```
{
                "config_entry_id": "ce7c454849da45978763dc8572dbe4f4",
                "device_id": "4635e97fdaef437cba7a4b96428a0ebd",
                "disabled_by": null,
                "entity_id": "sensor.grocy_stock",
                "name": null,
                "platform": "grocy",
                "unique_id": "202481593014205-stock"
            },
            {
                "config_entry_id": "ce7c454849da45978763dc8572dbe4f4",
                "device_id": "4ded9881e7aa4c85bb8401c2080fec38",
                "disabled_by": null,
                "entity_id": "sensor.grocy_chores",
                "name": null,
                "platform": "grocy",
                "unique_id": "202481593014205-chores"
            },
            {
                "config_entry_id": "ce7c454849da45978763dc8572dbe4f4",
                "device_id": "d1b203c7e88e482fb3da9e71b0014452",
                "disabled_by": null,
                "entity_id": "sensor.grocy_shopping_list",
                "name": null,
                "platform": "grocy",
                "unique_id": "202481593014205-shopping_list"
            },
            {
                "config_entry_id": "ce7c454849da45978763dc8572dbe4f4",
                "device_id": "227c181874af4dea90531764921a63b5",
                "disabled_by": null,
                "entity_id": "binary_sensor.grocy_expiring_products",
                "name": null,
                "platform": "grocy",
                "unique_id": "202481593014205-expiring_products"
            },
            {
                "config_entry_id": "ce7c454849da45978763dc8572dbe4f4",
                "device_id": "7a7606a25c22419883c71c8347526121",
                "disabled_by": null,
                "entity_id": "binary_sensor.grocy_expired_products",
                "name": null,
                "platform": "grocy",
                "unique_id": "202481593014205-expired_products"
            },
            {
                "config_entry_id": "ce7c454849da45978763dc8572dbe4f4",
                "device_id": "5bb83cd1479d40ce9a3633b7054d25f4",
                "disabled_by": null,
                "entity_id": "binary_sensor.grocy_missing_products",
                "name": null,
                "platform": "grocy",
                "unique_id": "202481593014205-missing_products"
            }
```

Sorry I don't understand "unique for Grocy instance"
You are right get_mac is unique on device running HA.

**Comment by SebRut** [comment]
Sorry for that late reply. if I understand correctly, the unique id should identify one entity uniquely. In our case the id should identify a sensor and it's corresponding grocy instance. This means 2 chores sensors for different grocy instances should have different ids but the id of one specific sensor should never change.

**Comment by BlueBlueBlob** [comment]
Like [sensor name] + [ sensor type] + get_mac() ?

**Comment by SebRut** [comment]
That would still lead to 2 different grocy instances in the same HA instance having the same id, as the mac is HA host dependent.

**Comment by BlueBlueBlob** [comment]
Ok sorry I just understand : it needs to work with multiple grocy instances.
We can add API key. Unique for each grocy instance. Not secure, but unique id is stored in internal files.

**Comment by SebRut** [comment]
hashed api key (maybe in combination with uri) sounds like the best we can up with atm, as the api doesn't provide a way of identifying itselft.

**Comment by BlueBlueBlob** [comment]
URL or URI (but I guess it's the URL and uri can be an URL)

SomeHash(API-Key + url ) ?

**Comment by SebRut** [comment]
I guess md5 of api key should be fine

File: custom_components/grocy/__init__.py [hunk 8]
```diff
@@ -4,6 +4,7 @@
 import logging
 import os
 from datetime import timedelta
+import hashlib
```
Comment (SebRut) [inline custom_components/grocy/__init__.py]: could you run `isort` over all files to ensure proper imports?

File: custom_components/grocy/__init__.py [hunk 9]
```diff
@@ -92,36 +115,26 @@
...
-            if not entry_config[CONF_ENABLED]:
-                continue
-
-            hass.async_create_task(
-                discovery.async_load_platform(
-                    hass, platform, DOMAIN, entry_config, config
-                )
-            )
+    hass.data[DOMAIN_DATA]["hash_key"] = hash_key
+    hass.data[DOMAIN_DATA]["url"] = url
```
Comment (SebRut) [inline custom_components/grocy/__init__.py]: The url is probably not needed for unique identification, could you either remove it from unique_id or hash the combination of api key and url?

File: custom_components/grocy/binary_sensor.py [hunk 10]
```diff
@@ -13,41 +15,61 @@
...
-    def __init__(self, hass, config):
+    def __init__(self, hass, sensor_type):
         self.hass = hass
+        self.sensor_type = sensor_type
         self.attr = {}
         self._status = False
-        self._name = config.get("name", DEFAULT_NAME) + ".expiring_products"
+        self._url = self.hass.data[DOMAIN_DATA]["url"]
+        self._hash_key = self.hass.data[DOMAIN_DATA]["hash_key"]
+        self._unique_id = '{}-{}-{}'.format(self._hash_key , self._url, self.sensor_type)
```
Comment (SebRut) [inline custom_components/grocy/binary_sensor.py]: see comment  in `__init__.py`

File: custom_components/grocy/sensor.py [hunk 11]
```diff
@@ -1,125 +1,103 @@
...
-    def __init__(self, hass, config):
+    def __init__(self, hass, sensor_type):
         self.hass = hass
+        self.sensor_type = sensor_type
         self.attr = {}
         self._state = None
-        self._name = '{}.products'.format(config.get("name", DEFAULT_NAME))
+        self._url = self.hass.data[DOMAIN_DATA]["url"]
+        self._hash_key = self.hass.data[DOMAIN_DATA]["hash_key"]
+        self._unique_id = '{}-{}-{}'.format(self._hash_key , self._url, self.sensor_type)
```
Comment (SebRut) [inline custom_components/grocy/sensor.py]: see comment in `__init__.py`

File: custom_components/grocy/__init__.py (hunk 8 above)
Comment (BlueBlueBlob) [reply custom_components/grocy/__init__.py]: Lol I didn't it was important. Thx
